### DISCUSSION TITLE
How to Dispose the MediaElement in .NET 8?

### DISCUSSION THREAD

**Discussion by marco-skizza** [CommunityToolkit/Maui#1663] [Q&A] [answered]
Hi

As it seems, the `Unloaded` event of a page was wrongly implemented in .NET 7 - where it was called when the page was "destroyed". This seems to be fixed now in .NET 8 and the `Unloaded` event gets called whenever the page exits the visual tree.

See also:
https://github.com/dotnet/maui/issues/16697

In .NET 8 and iOS I now have the problem, that this `Unloaded` event even gets called when entering the fullscreen mode in the `MediaElement`. So disposing the resources in this event is no option anymore.

See also the docu of `MediaElement`:
https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/views/mediaelement#clean-up-mediaelement-resources

So what's the recommended way to clean up the resources of a `MediaElement` in .NET 8?

**Comment by pictos** [comment]
@marco-skizza you can try to override the `OnNavigatedFrom` method, hopefully going to full screen mode will not trigger the Navigated event

**Reply by marco-skizza** [reply]
Thanks for your reply @pictos.

Unfortunately that doesn't work in my case. I have a "Next" page I can navigate to. So the `MediaElement` would be disposed when leaving the page - and when navigating back to the original page the `MediaElement` would already be disposed.

For me, the cleanest solution would be that the MAUI framework itself would provide us with some mechanism to call code when a page gets disposed. Like `Dispose`, `OnDisposing`, `OnDestroying`or something similar. But I dont't know if that's even possible...

What's your thought on this?

**Reply by pictos** [reply]
Yeah, that's more specific then... Well, there are a lot of ways to handle that, you can try to check the `HandlerChanged` event on your Page, when it becomes null (if will) then you can call the Disconnect, or think of a way to do it with PlatformBehavior, that will be Dettached when the Handler is cleaned up.

If you come with a sample project, that reproduces the issue it will open possibilities for more people to look at it

**Reply by marco-skizza** [reply]
Thanks @pictos

I now did the following for iOS:

```cs
    protected override void OnHandlerChanging(HandlerChangingEventArgs args)
    {
        base.OnHandlerChanging(args);

        if (args.NewHandler != null)
        {
            return;
        }

        mediaElement.Handler?.DisconnectHandler();
    }
```

This seems to work as expected on iOS.

However, it doesn't work on Windows as the method never gets called with `NewHandler` being `null`.
Unfortunately I couldn't test it on Android...

**Reply by pictos** [reply]
@PureWeen, this is the expected behavior for Windows? This the `OnHandlerChanging` of the `Page`

**Reply by PureWeen** [reply]
I'm surprised it works on any platform. We don't actively disconnect any handlers. Some stuff based on compatibility renderers might

**Reply by pictos** [reply]
@PureWeen I mean, the fact that on WIndows the handler is never `null` when the `OnHandlerChanging` is fired

**Reply by marco-skizza** [reply, marked as answer]
Hi

An updated answer for `.NET 9` and `CommunityToolkit.Maui.MediaElement` version `4.1.2`.
At it seems, you can now just call the following for `iOS`, `Android` and `Windows`:

```cs
    protected override void OnHandlerChanging(HandlerChangingEventArgs args)
    {
        base.OnHandlerChanging(args);

        if (args.NewHandler != null)
        {
            return;
        }

        MyMediaElement.Handler?.DisconnectHandler();
    }
```

See also my sample project [here](https://github.com/marco-skizza/CommunityMauiMediaElement).

**Reply by pictos** [reply]
@marco-skizza we just released a new version of Media element, can you try that and see if it's better?

**Reply by marco-skizza** [reply]
@pictos 

Yes, it works!
The memory leak is gone with the new version, when disposing the `MediaElement` as shown above.
Many thanks for all your effort!

**Reply by bzd3y** [reply]
I know this was closed, but I think there is more going on here, at least on Android. This may have to do with Maui in general, and not the `MediaElement`. But with that being said, I am still unclear on how to properly dispose the `MediaElement`.

In my situation, `Unloaded` is not being called on "leaving" the page, but that may be because I have pages being pushed to the navigation stack. But in that case, if I push the next page on the navigation stack and then return to the previous page that contained the `MediaElement` that is when `Unloaded` gets called, but it gets called *twice*. The first call disposes the `MediaElement` as above, and then the second call does nothing because it is already disposed.

But it gets even weirder, because in this case the "back" feature of app I maintain doesn't pop from the navigation stack. For reasons that aren't relevant here (though I am looking into changing this behavior anyway), the app recreates the previous page and pushes that onto the navigation stack (the app maintains its own stack of the content that has been displayed and uses that to go back).

And so considering that, I am confused as to why `Unloaded` was called at all, because the previous loaded view should still be on the navigation stack and continue to be "loaded", shouldn't it?

When the previous page is recreated, at least one of the `Unloaded` calls (I believe the first) is being called on the newly pushed view and so the `MediaElement` on that view has been disposed and is non-functional.

I'm guessing this is all due to Maui reusing views to optimize things or something like that. But if that is the case, then `Unloaded` seems like it cannot be relied on to dispose a `MediaElement` like this.

Further, from what I can tell, `OnHandlerChanging` doesn't get called with a null `NewHandler` at all during this process on Android, like it does on iOS, even after popping everything off of the navigation stack or manually disconnecting the handler like in the above answer.

**Reply by jo-ruch** [reply]
@marco-skizza I am still facing the same issue on Windows where my pages are not garbage collected (and all other issues you outlined in https://github.com/CommunityToolkit/Maui/issues/1702#issuecomment-1956125231) when using the suggested method of calling `MediaElement.Dispose()`.

Does it matter when this method is called? Has this issue really been resolved?

**Reply by pictos** [reply]
@jo-ruch @marco-skizza , are you forcing a GC collection by calling `GC.Collect` method? If even with that the control doesn't leak, please open an issue with a small project that reproduces the issue and tag me there.

**Reply by jo-ruch** [reply]
Hi @pictos , thank you for the quick reply. 

I suspect it is related to this issue: https://github.com/CommunityToolkit/Maui/pull/2265 where the dispose method on Windows does not clean up properly. I will open a new issue if this doesn't fix what I am experiencing.

**Reply by marco-skizza** [reply]
Hi @jo-ruch, Hi @pictos
Thanks for bringing this up again. I did some testing with `.NET 9` and the current version of `MediaElement`.
I updated the answer above. And yes, `MyMediaElement.Dispose()` doesn't seem to be necessary anymore.

**Comment by CliffAgius** [comment]
Closed as answered
